library/psa_crypto.c

```diff
@@ -925,17 +972,53 @@ psa_status_t psa_destroy_key( psa_key_handle_t handle )
     psa_key_slot_t *slot;
     psa_status_t status = PSA_SUCCESS;
     psa_status_t storage_status = PSA_SUCCESS;
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+    psa_se_drv_table_entry_t *driver;
+#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
 
     status = psa_get_key_slot( handle, &slot );
     if( status != PSA_SUCCESS )
         return( status );
+
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+    driver = psa_get_se_driver_entry( slot->lifetime );
+    if( driver != NULL )
+    {
+        psa_crypto_prepare_transaction( PSA_CRYPTO_TRANSACTION_DESTROY_KEY );
+        psa_crypto_transaction.key.lifetime = slot->lifetime;
+        psa_crypto_transaction.key.slot = slot->data.se.slot_number;
+        psa_crypto_transaction.key.id = slot->persistent_storage_id;
+        status = psa_crypto_save_transaction( );
+        if( status != PSA_SUCCESS )
+        {
+            /* TOnogrepDO: destroy what can be destroyed anyway */
+            return( status );
+        }
+
+        status = psa_destroy_se_key( driver, slot->data.se.slot_number );
+    }
+#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
+
 #if defined(MBEDTLS_PSA_CRYPTO_STORAGE_C)
     if( slot->lifetime == PSA_KEY_LIFETIME_PERSISTENT )
     {
         storage_status =
             psa_destroy_persistent_key( slot->persistent_storage_id );
     }
 #endif /* defined(MBEDTLS_PSA_CRYPTO_STORAGE_C) */
+
+#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
+    if( driver != NULL )
+    {
+        status = psa_crypto_stop_transaction( );
+        if( status != PSA_SUCCESS )
+        {
+            /* TOnogrepDO: destroy what can be destroyed anyway */
+            return( status );
+        }
+    }
+#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
```

Comment:
I'll add a comment, but maybe part of the confusion is that you seem to expect that `psa_crypto_save_transaction` and `psa_crypto_stop_transaction` are somehow redundant. Actually `save` starts the transaction and `stop` marks it as finished. Do you think the function names should be changed?

Reply:
Doesn't stop remove the transaction from storage? If so, then I'm not sure how the transaction is committed before it disappears.

Given your prose description, we should use terms that match as that's probably the more intuitive way to think about it. This will match the shape of the rest of the API as well. psa_crypto_start_transaction() and psa_crypto_finish_transaction().

Reply:
There's no function that I'd really want to call `start`. The first function `prepare`, which does not save anything to storage. Should it be called `start`, to be paired with `stop`? But `stop` _does_ access storage. I find `finish` less descriptive than `stop` for what it does, because I'd expect `finish` to do some kind of committing, but `stop` doesn't, it just removes the information about a transaction in progress.

Reply:
OK, I understand this mechanism now. The "committing" is happening in `psa_destroy_se_key()` in this case. We are saying "I'm going to do this." and the we do it, and then we say "I finished doing this."

In crypto init, if we have a pending transaction, we have enough information to complete the pending transaction (or check that it was done already). I'm not sure when we'd choose to roll back the transaction, though, nor what operation we'd perform to do the rollback (like adding the key back to the key slot for example, since the SE never completed removal, maybe).

Reply:
Rolling back a key destruction is not viable - we cannot determine if the key material is or is not present in the SE and recover it if not. But as this API is destructive, it is easy to record enough information to retry/complete the key destruction.

For key creation/import/derivation, adopting a rolling-back approach means that the transaction does not need to retain any key material, only the key meta-data. On restart, an incomplete transaction can be rolled back by deleting any maybe-created key material. Adding an extra save of the transaction entry after the SE has created the material to record that fact before making updates to the driver's persistent data and the core's key meta-data records would enable a roll-forward as the secrets have been populated in the SE.

Reply:
... but that adds an extra transaction update to benefit a rare situation (system fail during persistent SE key creation).

Reply:
> Rolling back a key destruction is not viable

it depends on the order of operations, doesn't it? If we start the transaction, remove the information from core, and then fail before removing the information from the SE, we could put the information back into the core given information in the transaction. If we really can't determine if the key is partially removed or still present in the SE, then I agree, even with this order of operations, we can't roll back the destruction as we need a way to validate that the key is still in the SE (as we could have failed/lost power after SE removal, but before transaction completion).

Not necessary for this PR, but we should write down what our policies should be (at least for our implementation, not necessarily anything at the spec level) for when we roll-back and when we automatically retry a transaction (if ever). We then can test against our written policy with mocks and fault injection to ensure we adhere to our policy.